crates/file/src/file_reader.rs

```diff
@@ -0,0 +1,268 @@
+        }
+    }
+
+    /// Read the contents of `blob` into an array buffer, and then run the callback.
+    ///
+    /// If the returned `FileReader` is dropped before the callback is invoked, the read will be
+    /// cancelled.
+    pub fn read_as_array_buffer<F>(blob: &Blob, callback: F) -> FileReader
+    where
+        F: FnOnce(Result<js_sys::ArrayBuffer, FileReadError>) + 'static,
+    {
+        let load_callback: Arc<
+            Mutex<Option<Box<dyn FnOnce(Result<js_sys::ArrayBuffer, FileReadError>) + 'static>>>,
+        > = Arc::new(Mutex::new(Some(Box::new(callback))));
+        let error_callback = load_callback.clone();
+
+        let reader = web_sys::FileReader::new().unwrap_throw();
+
+        // Either the load listener or the error listener will be called, never both (so FnOnce is
+        // ok).
+        let load_listener = EventListener::new(&reader, "load", move |event| {
+            let reader = reader_from_event(&event);
+            let result = reader
+                .result()
+                .unwrap_throw()
+                .dyn_into::<js_sys::ArrayBuffer>()
+                .unwrap_throw();
+            load_callback.lock().unwrap_throw().take().unwrap_throw()(Ok(result));
+        });
+
+        let error_listener = EventListener::new(&reader, "error", move |event| {
+            let error = error_from_event(&event);
+            error_callback.lock().unwrap_throw().take().unwrap_throw()(Err(error));
+        });
+
+        reader.read_as_array_buffer(blob.as_ref()).unwrap_throw();
+        FileReader {
+            reader,
+            load_listener,
+            error_listener,
+        }
+    }
+
+    /// Read the contents of `blob` into a `Vec<u8>`, and then run the callback.
+    ///
+    /// If the returned `FileReader` is dropped before the callback is invoked, the read will be
+    /// cancelled.
+    pub fn read_as_bytes<F>(blob: &Blob, callback: F) -> FileReader
+    where
+        F: FnOnce(Result<Vec<u8>, FileReadError>) + 'static,
+    {
+        read_as_array_buffer(blob, move |result| {
+            callback(result.map(|buffer| js_sys::Uint8Array::new(&buffer).to_vec()));
+        })
+    }
+
+    /// Get the event from the reader target, doing necessary casting.
+    fn reader_from_event(event: &web_sys::Event) -> web_sys::FileReader {
+        let target = event.target().unwrap_throw();
+        debug_assert!(target.is_instance_of::<web_sys::FileReader>());
+        target.unchecked_into()
+    }
+
+    fn error_from_event(event: &web_sys::Event) -> FileReadError {
+        let exception = reader_from_event(event).error().unwrap_throw();
+        match exception.name().as_str() {
+            "NotFoundError" => FileReadError::NotFound(exception.message()),
+            "NotReadableError" => FileReadError::NotReadable(exception.message()),
+            "SecurityError" => FileReadError::Security(exception.message()),
+            name => throw_str(&format!("unrecognised error type: {}", name)),
```

Comment:
I'd really like to avoid the formatting code if possible, but I'm not sure what way would be best.

Reply:
Do you mean to avoid pulling in `std::fmt` so it can be pruned if no user code uses it?

Reply:
I mean that it shouldn't be using `format!`, since that pulls in all of the formatting infrastructure.

Implementing `std::fmt::Display` or `std::fmt::Debug` is fine, since Rust will remove it if it's not used.

It's the act of *using* `std::fmt::Display` or `std::fmt::Debug` that pulls it in. Normally that's done by using `format!` or `println!`.

Reply:
In that case we also want to exclude the `write!` macro used in `impl Display for FileReadError`.